### ISSUE TITLE
Universal specs for JSON-RPC API methods

### ISSUE THREAD

**Issue by cdetrio** [ethereum/EIPs#217]
## Universal specs for JSON-RPC API methods

This is a pre-draft outline of a [Hive](https://github.com/karalabe/hive)-based multi-client test suite for the [JSON-RPC API](https://github.com/ethereum/wiki/wiki/JSON-RPC). The specification format is [JSON schema](http://json-schema.org/) (inspired by the [Kodi](http://kodi.wiki/view/JSON-RPC_API/v4) API description).

Existing RPC test suites include [ethereum/rpc-tests](https://github.com/ethereum/rpc-tests) and [bas-vk/hive/tree/rpc](https://github.com/bas-vk/hive/tree/rpc/simulators/ethereum/rpc/eth). The former is based on mocha.js and the latter on go-ethereum's `ethclient.Client` api. This proposal intends to be a language and client-agnostic test and specification format.

Here's an example JSON schema specification, for [eth_getBlockTransactionCountByHash](https://github.com/cdetrio/interfaces/blob/master/rpc-specs-tests/schemas/eth_getBlockTransactionCountByHash.json), annotated for explanation:

```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "eth_getBlockTransactionCountByHash",
    "title": "eth_getBlockTransactionCountByHash",
    "description": "eth_getBlockTransactionCountByHash JSON-RPC method request and response schema.",

    "request": {
      "id": "#request",
      "allOf": [ // allOf means the request must pass both sub-schemas
        { "$ref": "jsonrpc-request.json" }, // first-pass generic validation, see file https://github.com/cdetrio/interfaces/blob/master/rpc-specs-tests/schemas/jsonrpc-request.json
        { "$ref": "#/definitions/request-obj" } // second-pass validation, specific to the RPC method
      ],
      "definitions": {
        "request-obj": {
          "id": "#request-obj",
          "properties": {
            "method": {
              "type": "string",
              "enum": ["eth_getBlockTransactionCountByHash"]
            },
            "params": {
              "type": "array",
              "items": [
                {
                  "type": "string",
                  "description": "DATA, 32 Bytes - Hash of a block."
                }
              ]
            }
          }
        }
      }
    },

    "response": {
      "id": "#response",
      "allOf": [
        { "$ref": "jsonrpc-response.json" },
        { "$ref": "#/definitions/response-obj" }
      ],
      "definitions": {
        "response-obj": {
          "id": "#response-obj",
          "properties": {
        		"result": {
              "type": "string",
              "description": "QUANTITY - integer of the number of transactions in this block."
        		}
        	}
        }
      }
    }
}
```

The `description` fields are not used for validation, but for generating RPC API method documentation as seen on the [wiki](https://github.com/ethereum/wiki/wiki/JSON-RPC).

The JSON schema specifications are paired with test cases. Here's an example file with test cases for [eth_getBlockTransactionCountByHash](https://github.com/cdetrio/interfaces/blob/master/rpc-specs-tests/tests/eth_getBlockTransactionCountByHash.json):

```
{
  "title" : "eth_getBlockTransactionCountByHash",

  "schema": {
    "$ref": "../schemas/eth_getBlockTransactionCountByHash.json"
  },

  "chainConfig" : {
    "$ref": "../configs/bcRPC_API_Test.json"
  },

  "tests": [
    {
      "title": "eth_getBlockTransactionCountByHash for block with one tx",
      "request" : {
        "method" : "eth_getBlockTransactionCountByHash",
        "params" : ["0x4e9a67b663f9abe03e7e9fd5452c9497998337077122f44ee78a466f6a7358de"]
      },
      "expectedResponse" : {
        "result": "0x1"
      },
      "asserts": [
        {
          "description": "response is not empty",
          "program": ".receivedResponse.result != null"
        },
        {
          "description" : "transaction count should be equal",
          "program" : ".receivedResponse.result == .expectedResponse.result"
        }
      ]
    }
  ]
}
```

The `asserts` array specifies [jq programs](https://stedolan.github.io/jq/manual/) or "filters", which are used to compare the received response against the expected response.

The `chainConfig` option points to a set of blocks/transactions to import. The current chain configuration is inherited from [ethereum/rpc-tests](https://github.com/ethereum/rpc-tests/tree/master/lib), which in turn borrows from the [state tests](https://github.com/ethereum/tests/blob/25912e023e7cf25c33ed6dff078df0c941f2c7d6/BlockchainTests/bcRPC_API_Test.json) repo. This configuration file is loaded by the [test runner](https://github.com/cdetrio/hive/blob/master/simulators/ethereum/rpeecee/simulator.py#L52-L58), and [imported by](https://github.com/cdetrio/hive/blob/master/clients/go-ethereum:master/geth.sh#L82-L94) the hive clients.

After hive runs the RPC tests against a set of clients, the results are displayed as a compatibility table: [http://cdetrio.github.io/eth-compat-table/](http://cdetrio.github.io/eth-compat-table/) (inspired by the table for [ES6](https://kangax.github.io/compat-table/es6/)).

## Todos

One remaining task is to specify a test format for the [pub/sub](https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB) methods. One possibility is a `transitionEvent` field to specify an event (e.g. arrival of a block or a pending transaction), and an `onTransition` field to describe the expected notifications.

## Test suite maintenance

An automated maintenance process for the RPC test suite will be described in a forthcoming Meta-EIP. The current plan is to maintain updates to the RPC test suite through a [voting contract](https://github.com/MetaMask/IPFS-Ethereum-Hackathon/issues/14), with one-developer one-vote where developer identities are verified using [devcon2 tokens](https://www.devcon2-token.com/). (There is a separate plan to expand the Devcon2 tokens into a web of trust, in order to enable participation by any and all Ethereum developers/stakeholders). Tentatively named `EipSignal`, this automated governance process is somewhat in the spirit of the [Yellow Paper Committee](https://github.com/gavofyork/curly-engine); however, the outcome of `EipSignal` votes will only mandate changes to the RPC API specification and test suite as described above, and will NOT govern "core/consensus" protocol changes. The work-in-progress on `EipSignal` can seen here: [https://github.com/MetaMask/IPFS-Ethereum-Hackathon/issues/14](https://github.com/MetaMask/IPFS-Ethereum-Hackathon/issues/14).

**Comment by axic** [comment]
👍 

> The asserts array specifies jq programs or "filters", which are used to compare the received response against the expected response.

Is it possible to include strict verifications, such as a subset of regular expressions on some data fields? Thinking here about validating that an actual address was returned for example by `personal_newAccount`.

**Comment by flyswatter** [comment]
This is very exciting to me. Developers need a place they can look up stable methods, and client devs need a smooth way to evolve their interfaces.

Other clients that will be nice to add to the tested columns:

- [testrpc](https://github.com/ethereumjs/testrpc)
- [provider-engine](https://github.com/MetaMask/provider-engine) (Wrapped with an RPC server, backed by different clients)

And I'm sure others, too. Once we have a strong singular test suite like this, building new RPC providers should become a much more stable process.

**Comment by bobsummerwill** [comment]
Really great to see this happening, @cdetrio, @flyswatter!

Starting with the RPCs makes sense, and I hope that the patterns which are built can be extended to broader specifications over time.   We've had discussions along very similar lines for Enterprise using Boardroom.

See also http://html5test.com.

**Comment by cdetrio** [comment]
> Is it possible to include strict verifications, such as a subset of regular expressions on some data fields?

Yes, actually there's two ways to enforce a regexp. It can be done in the JSON schema using [`pattern`](http://json-schema.org/latest/json-schema-validation.html#rfc.section.3.3), or as a test case using a [jq filter](https://stedolan.github.io/jq/manual/#RegularexpressionsPCRE).

**Comment by cdetrio** [comment]
> Have you seen cpp-ethereum's test_setChainParams RPC method? I think that might be the successor of that format.

Ah, here's an issue on that: https://github.com/ethereum/interfaces/issues/4

**Comment by cdetrio** [comment]
There's also https://github.com/ethcore/ethereum-rpc-json, which generates documentation but doesn't include tests to run against clients.

**Comment by rphmeier** [comment]
> http://cdetrio.github.io/eth-compat-table/ 
Seems like it measures compatibility against Geth, not against any spec.

**Comment by bobsummerwill** [comment]
Hey @rphmeier.

See "Test suite maintenance" at the very top of this issue.

The existing tests are incomplete placeholders.

**An automated maintenance process for the RPC test suite will be described in a forthcoming Meta-EIP. The current plan is to maintain updates to the RPC test suite through a voting contract, with one-developer one-vote where developer identities are verified using devcon2 tokens. (There is a separate plan to expand the Devcon2 tokens into a web of trust, in order to enable participation by any and all Ethereum developers/stakeholders). Tentatively named EipSignal, this automated governance process is somewhat in the spirit of the Yellow Paper Committee; however, the outcome of EipSignal votes will only mandate changes to the RPC API specification and test suite as described above, and will NOT govern "core/consensus" protocol changes. The work-in-progress on EipSignal can seen here: MetaMask/IPFS-Ethereum-Hackathon#14.**

**Comment by flyswatter** [comment]
@bobsummerwill beat me to it, but I'll still reword:

@rphmeier 
> Seems like it measures compatibility against Geth, not against any spec.

There is no current spec to measure against, which is why this EIP exists.

The current plan is to [manage this test suite with a form of governance](https://github.com/MetaMask/IPFS-Ethereum-Hackathon/issues/14), at which point non-geth-compliant specs may emerge, or even become dominant.

Until then, the linked test suite should be treated as a template for specs, not as a proposed spec itself.

**Comment by rphmeier** [comment]
A test suite isn't a specification, it just tests conformity to some specification. The schema described here is not sufficient to specify certain primitives which all methods rely upon. This is exactly how you get into the situation of strange corner cases that each client implements differently while still being "to-spec".

The places where I see the current documentation lacking the most:
  - definition of pending state
  - definition of error messages
  - lack of consensus between client devs over changes

I'm glad that this EIP is working on point 3. But we need strong definitions of the primitives used to build the RPC methods rather than specific test cases which will just lead to overfitting.

It's misleading to label one specific implementation of ambiguous methods as "correct".

**Comment by flyswatter** [comment]
@rphmeier 

I think we're actually starting to stray beyond the scope of this issue, but there isn't a better place for it.  This issue is really about an RPC spec format.  We're beginning to talk about the governance process of managing that spec.

I understand the initial alarm, it could sound like someone is trying to hijack what "correct" means, but that's not our intention at all.  We're not trying to shame Parity over a specific difference from Geth, we're trying to create a tool for all client devs to better coordinate their implementations.

The (current, in flux) plan is that a given suite of tests would *accompany* an EIP spec, not replace it. For example, for a given EIP, a variety of accompanying test suites could be nominated, and the most popular one(s?) could be displayed on the compatibility table.

Since the most popular test suite for a given EIP would be displayed, the test would not itself be an endorsement of that EIP spec, it would merely be a representation of it, and a way of automatically measuring compliance to it, which better represents the sort of ad-hoc standards process we actually have, where endorsement is almost solely represented through client implementation.

It certainly would not be a spec itself, since the table could include distinct, incompatible implementations of the same method, but it *would* provide a singular place to view and compare those incompatibilities. This table *would not be about labeling any one implementation as correct*, but instead, be a way of highlighting differences of implementation, for the benefit of both dapp and client developers.

I don't think a test suite has to encourage over-fitting, either.  In the test format that @cdetrio has proposed, specific values of the response can be used to validate results, so it does not need to blindly enforce bit-conformity.

In the case of `personal_sign`, a reasonable test might check if a certain account with a published private key, given a certain message, produces a consistent signed output, and no more.  This is not a spec, but adhering to it, and publishing this sample data, could have alleviated many headaches that came from having an under-specified feature adopted by a client.

I'm very interested in seeing solutions to 1 & 2 as well, and while a test suite may not be a complete solution for them, I think it may still be a useful tool for even those issues, when combined with EIP specifications.

**Comment by cdetrio** [comment]
@rphmeier Good points, I agree that pending state and error messages are two important areas to address.

EIP https://github.com/ethereum/EIPs/issues/136 addresses RPC error messages, and I think the schema here is sufficient to incorporate those error messages.

Pending state is trickier. Do you have any suggestions for how to define the pending state as a primitive?

I'm still unsure about this, but one thought is that perhaps an additional [testing-specific](https://github.com/ethereum/interfaces/issues/4) RPC method would be useful, such as a `test_setPendingBlock(block_rlp)` method, similar to the `test_addBlock(block_rlp)` already supported by cpp-ethereum. Such RPC methods would make it easy to test assertions in a `transitionEvent` field (mentioned in the Todos above). On the other hand, I dislike the idea of test-specific RPC methods because it could lead to clients implementing the correct behavior for the test methods, but unexpected behavior in normal use. An example of this is https://github.com/ethereum/go-ethereum/issues/2897, where `eth_sendTransaction` applies tx's to the pending block only if the client is not mining (if the client is mining, then you have to wait a block before the tx is applied to the pending state). Preventing these kinds of corner cases is the goal of black-box testing using Hive (as opposed to unit tests). I'd prefer to just use `eth_sendTransaction` and assert that the tx is applied immediately, specifying it as incorrect behavior if not, regardless of whether the client is mining.

On the third point, "lack of consensus between client devs over changes", that will be addressed by a separate EIP/meta-EIP, which will propose a governance/maintenance process. However, I wouldn't mind bootstrapping it with an initial test suite of "compatible/correct" behavior, determined informally by pre-existing standards and rough consensus. The current examples were generated against geth (two incompatible geth branches, in fact), but they are certainly only meant as examples.

The discussion here should focus on the spec or test schema/format. The term "spec" is used loosely, in the sense of an [API spec](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md), where the aim is to enable black-box testing of RPC method behavior. We want to settle on a format that is sufficient to specify behavior (at least implicitly if not explicitly) that is currently ambiguous, and test compliance.

**Comment by MicahZoltu** [comment]
What I see as a major hole in these tests is that they don't appear to have a mechanism for setting up starting state.  What these tests really should be testing is along the lines of:
> Given state A; call RPC method B then C; expect result D

This means tests necessarily need to be more complex than a simple "given request, expect result".  One way to achieve this (pseudocode only):
```
initializeGenesisBlock()
transactionHash = submitTransaction(myTransaction)
expect(transactionHash).to.not.be.null
mineAllTransactions()
newBlock = rpc_eth_getBlockByNumber("latest")
expect(newBlock.transactions).to.contain(transactionHash)
```

Alternatively, the test could explicitly setup expected state.  For example:
```
startingState: [
	accounts: [
		{ address: "0x1234abcd", eth: "5", signingStrategy: "SIGN_EVERYTHING" },
	],
	blocks: [
		{ transactions: [], uncles: [], hash: "0xabcd1234", ... },
		{ transactions: [...], uncles: [...], hash: "...", ... },
	],
	contractState: [
		...
	],
	...
]
```
Then the test would call a one or more RPC methods and assert that it gets back a very well defined response.

The key here is that all tests should setup the current full chain state, then call a particular RPC method and assert on the results.  Building a test suite like this is expensive, but is also valuable for multi-implementation compatibility.  Note that this test doesn't care _how_ an implementation achieves the result (implementation details), but it is well defined that given a particular state and an RPC call, an expected outcome is achieved.

As to how to codify "setup", I'll leave that up for discussion but I do think it is critical that __COMPLETE__ setup is included in the test.  The test should assume nothing about the current state of the world when it executes.

Personally, I'm a fan of the first style because it is generally easier to write tests when the test defines the path to the expected state, rather than requiring that the thing under test has the ability to set the current state.  The first mechanism also allows true black box testing, where the test can be run against an endpoint without knowing its implementation.  The second mechanism requires that the test have a back-door mechanism to setup current state.

**Comment by LogvinovLeon** [comment]
Hi All,

Leo from 0xproject here.

Ater reading this EIP I was disappointed by how people tried to not understand @MicahZoltu.
I agree with him. Signing messages is as important as signing transactions and users that sign anything without checking, but transactions are those - whom we can't help.

More and more value transfer will happen not (or only partially) as a result of a transaction.
0x protocol uses signing to sign off-chain orders, which can later be plugged into the smart contract.

With the current implementation I have two options.
Either sign using `eth_sign` and getting a warning on MetaMask.
<img width="359" alt="screen shot 2017-09-07 at 13 34 05" src="https://user-images.githubusercontent.com/6204356/30161537-44a043b6-93d1-11e7-989a-d0f84da17144.png">

Sign using `personal_sign` and get a MetaMask popup with unreadable symbols, cause Metamask assumed, I'm signing ASCII, when I was signing bytes.
<img width="361" alt="screen shot 2017-09-07 at 13 27 50" src="https://user-images.githubusercontent.com/6204356/30161571-685dca08-93d1-11e7-852f-d7e1ff4a920e.png">

In the fist case - users can at least verify the hash and be sure - they're secure. In the second case - there is nothing they can do.

I'm a fan of a solution with prefixes for all possible data types:
`sign_bytes`, `sign_ascii`, `sign_number`, `sign_utf8`

**Comment by MicahZoltu** [comment]
@LogvinovLeon Did you mean to put this comment over on the `eth_sign` thread?

**Comment by holiman** [comment]
I just now stumled upon this when searching for something like [swagger](https://swagger.io/swagger-editor/) to define and generate documentation for our json-rpc apis. Unfortunately, swagger isn't for json-rpc. 

@cdetrio Have you found any good editors to write/generate the specifications above? Or did you do it by hand? 
I also found [jrgen](https://github.com/mzernetsch/jrgen), which is more geared towards our usecase, but doesn't appear quite as handy as the swagger editor. 

I'd really like to have not only a specification, but also an editor which makes writing/validating the specification intuitive and simple. If it involves writing files by hand, then running some npm script to validate and generate docs based on our custom specification file, then I'm afraid it will just bitrot after a while.

**Comment by winsvega** [comment]
Totally up for this. 
I could write the boost test cases and you could run retesteth tool against any clients to check request-response schemas.

**Comment by pipermerriam** [comment]
@holiman I think I'd like to pick this up on my team.  My thought was that we could add these schemas to the `ethereum/tests` repository and document how they should be run against the client's JSON-RPC responses.  It'd still be up to each client to implement the tests.

**Comment by bobsummerwill** [comment]
Do it, @pipermerriam.  That would be awesome!

When I spoke at the launch event for the EEA back in February 2017, 17 months ago, this was actually one of the areas of protocol specification which I highlighted as being most needed and highest impact - both for public Ethereum and for enterprise scenarios.

The lack of clear and testable JSON-RPC specifications really hurts everyone trying to build on top of Ethereum.

What are PegaSys doing for RPCs in Pantheon, @shahankhatch?   Are you just trying to be close to Geth?   Or did you do some modal thing like Parity have?

**Comment by bobsummerwill** [comment]
@rphmeier What do you see as the best path towards standardization of RPC specs across the various Ethereum clients, Robert?

As you said last year, there is a balance here to ensure that we don't "over-fit", and to ensure that any automated tests that we build are encoding unambiguous *original* specifications.

Any thoughts, @gavofyork?

**Comment by winsvega** [comment]
retesteth is capable of running response requests check against clients via RPC 
we just need to come up with default parameters that node will report to those commands. 
eg.  peercount 0.  accounts - empty   and so on.

**Comment by ethers** [comment]
Proposal: https://ethereum-magicians.org/t/eip-remote-procedure-call-specification/1537
